src/backend/distributed/metadata/metadata_cache.c

```diff
@@ -4748,24 +4754,61 @@ InvalidateForeignKeyGraph(void)
 }
 
 
+/*
+ * NeedsMetadataCacheInvalidation returns true if a given oid is cached in MetadataCache.
+ * This list does not include the Oids for which there are seperate invalidation
+ * functions. E.g. distColocationRelationId is handled in
+ * InvalidateForeignRelationGraphCacheCallback(Datum argument, Oid relationId).
+ * So we do not include it here.
+ */
+static bool
+NeedsMetadataCacheInvalidation(Oid relationId)
+{
+	if (relationId == MetadataCache.distPartitionRelationId ||
```

Comment:
I thought it was surprising that this was checking the table Oids instead of the index Oids, because the index oids are actually changed when a REINDEX CONCURRENTLY happens. 

So I had expected that an invalidation would be sent on the index its oid instead of the one for the table that it's an index for. As it turns out it's more interesting than that. Invalidations are sent for many things when a REINDEX cuncurrently happens:
1. The oid of the table
2. The oid of the old index
3. The oid of the new index

One thing I'm worried about with the current change is that we might flush the cache more often than before, maybe even much more often depending on the workload. My suggestions:
1. Let's keep the current behaviour in this PR, only flush on distPartitionRelationId invalidations.
2. Let's create a second PR that fixes it for all indexes.
3. I think it would be better if that PR would not listen for invalidations for table oid, but instead only for index oids. Then when a cached index gets an invalidation message, we should not invalidate the whole cache, but instead only set the matching `distXyzIndexId` field to `InvalidOid`.

PS. The minimal example of the original issue that I could find is by running a REINDEX CONCURRENTLY twice on `pg_dist_partition_logical_relid_index`, i.e.:
```
REINDEX INDEX CONCURRENTLY pg_catalog.pg_dist_partition_logical_relid_index;
REINDEX INDEX CONCURRENTLY pg_catalog.pg_dist_partition_logical_relid_index;
```